File: mod/bridge.lua
local BRIDGE_VERSION = 2
local last_version_warning = 0

local RECV_MOD_FS_NAME = "coop-voicechat-recv"
local FILE_HEADER = "smvc"

-- mod_fs_hide_errors(true)

gVoiceBridge = {}

gVoiceBridge.connected = false

gVoiceBridge.sendFs = mod_fs_get()
gVoiceBridge.recvFs = mod_fs_get(RECV_MOD_FS_NAME)

gVoiceBridge.sendFs:clear()

gVoiceBridge.syncFile = gVoiceBridge.sendFs:create_file("sync", false)
mod_fs_file_set_compression(gVoiceBridge.syncFile, 0)

gVoiceBridge.syncLocalFrame = 1
gVoiceBridge.syncRemoteFrame = 0
gVoiceBridge.syncRemoteAckFrame = 0
gVoiceBridge.syncLastRemoteFrame = 0
gVoiceBridge.syncTimeoutCounter = 0

local function bridge_connect()
    gVoiceBridge.connected = true
    play_sound(SOUND_GENERAL_COIN, gGlobalSoundSource)
    djui_popup_create("Voice Chat:\n\\#60f060\\Client Connected!", 2)
end

local function bridge_disconnect()
    gVoiceBridge.connected = false
    play_sound(SOUND_MENU_PAUSE_HIGHPRIO, gGlobalSoundSource)
    djui_popup_create("Voice Chat:\n\\#ffa060\\Client Disconnected!", 2)
end

-- checks if new data is available
local function bridge_poll()
    mod_fs_hide_errors(true)
    -- this is the most likely operation to fail
    gVoiceBridge.recvFs = mod_fs_reload(RECV_MOD_FS_NAME)
    mod_fs_hide_errors(false)
    if not (gVoiceBridge.sendFs and gVoiceBridge.recvFs) then
        return false
    end

    local syncFile = gVoiceBridge.recvFs:get_file("sync")
    if not syncFile then
        return false
    end

    syncFile:seek(4, FILE_SEEK_SET)

    local remoteVersion = syncFile:read_integer(INT_TYPE_U16)
    if remoteVersion ~= BRIDGE_VERSION then
        local next_version_warning = last_version_warning + 900
        if last_version_warning == 0 or get_global_timer() >= next_version_warning then
            last_version_warning = get_global_timer()
            djui_popup_create(
                string.format("\\#f06060\\Voice Chat version mismatch!\n\\#dcdcdc\\Mod version: %d\nClient version: %d",
                    BRIDGE_VERSION, remoteVersion), 3)
        end
        return false
    end

    local lastActive = gVoiceBridge.connected
    local lastRemoteFrame = gVoiceBridge.syncRemoteFrame

    gVoiceBridge.syncRemoteFrame = syncFile:read_integer(INT_TYPE_U32)
    gVoiceBridge.syncRemoteAckFrame = syncFile:read_integer(INT_TYPE_U32)

    local ackFrameValid = gVoiceBridge.syncRemoteAckFrame > 0 and
        gVoiceBridge.syncRemoteAckFrame <= gVoiceBridge.syncLocalFrame
    local ackFrameThreshold = gVoiceBridge.syncLocalFrame - gVoiceBridge.syncRemoteAckFrame < 6

    gVoiceBridge.syncLocalFrame = get_global_timer()
    gVoiceBridge.syncLastRemoteFrame = lastRemoteFrame

    if gVoiceBridge.syncRemoteFrame > lastRemoteFrame then
        -- active means the client is running and acknowledging us
        local shouldActivate = ackFrameValid and ackFrameThreshold
        if shouldActivate then
            gVoiceBridge.syncTimeoutCounter = 0
            if not lastActive then bridge_connect() end
        end
        return gVoiceBridge.connected
    end

    if lastActive and not (ackFrameValid and ackFrameThreshold) then
        gVoiceBridge.syncTimeoutCounter = gVoiceBridge.syncTimeoutCounter + 1
        if gVoiceBridge.syncTimeoutCounter > 6 then
            log_to_console(
                string.format("Bridge disconnected - av:%s aft:%s slf:%d srf:%d sraf:%d stc:%d", ackFrameValid,
                    ackFrameThreshold,
                    gVoiceBridge.syncLocalFrame, gVoiceBridge.syncRemoteFrame, gVoiceBridge.syncRemoteAckFrame,
                    gVoiceBridge.syncTimeoutCounter),
                CONSOLE_MESSAGE_WARNING)
            bridge_disconnect()
        end
    end

    return false
end

-- handles new data
local function bridge_recv()
    audio_recv()
end

-- sends new data
local function bridge_send()
    local f = mod_fs_get_or_create_file(gVoiceBridge.sendFs, "local_player", false)
    mod_fs_file_clear(f)
    f:write_bytes(FILE_HEADER)

    mod_fs_file_write_player(f, 0)

    f = mod_fs_get_or_create_file(gVoiceBridge.sendFs, "players", false)
    mod_fs_file_clear(f)
    f:write_bytes(FILE_HEADER)

    for i = 1, MAX_PLAYERS - 1 do
        local lVoiceState = gVoiceStates[i]
        local lNetworkPlayer = gNetworkPlayers[i]
        if lNetworkPlayer.connected then
            f:write_integer(i | 0x80, INT_TYPE_U8)
            f:write_integer(#lVoiceState.audioFile, INT_TYPE_U8)
            f:write_bytes(lVoiceState.audioFile)
            mod_fs_file_write_player(f, i)
        else
            f:write_integer(i, INT_TYPE_U8)
        end
    end

    audio_send()
end

local function bridge_update()
    if bridge_poll() then
        bridge_recv()
        bridge_send()
    end

    gVoiceBridge.syncFile:rewind()
    gVoiceBridge.syncFile:write_bytes(FILE_HEADER)
    gVoiceBridge.syncFile:write_integer(BRIDGE_VERSION, INT_TYPE_U16)
    gVoiceBridge.syncFile:write_integer(gVoiceBridge.syncLocalFrame, INT_TYPE_U32)
    gVoiceBridge.syncFile:write_integer(gVoiceBridge.syncRemoteFrame, INT_TYPE_U32)

    gVoiceBridge.sendFs:save()
end

hook_event(HOOK_UPDATE, bridge_update)

File: client/coop/bridge.go
package coop

import (
	"coop-voicechat/config"
	"log"
	"time"
)

const BRIDGE_VERSION uint16 = 1

const POLL_INTERVAL = 33           // 1/30
const POLL_INTERVAL_INACTIVE = 100 // 1/10

type Bridge struct {
	Running   bool
	Connected bool
	Event     chan BridgeEvent
	Options   *config.Config

	SendFs *ModFS
	RecvFs *ModFS

	syncLocalFrame      uint32
	syncRemoteFrame     uint32
	syncRemoteAckFrame  uint32
	syncLastRemoteFrame uint32
	syncTimeoutCounter  uint8
}

func NewBridge(options *config.Config) *Bridge {
	send_modfs, err := ModFsGet("coop-voicechat-recv")
	if err != nil {
		panic(err)
	}
	send_modfs.Write()

	recv_modfs, err := ModFsGet("coop-voicechat")
	if err != nil {
		panic(err)
	}
	recv_modfs.Write()

	return &Bridge{
		Running:   false,
		Connected: false,
		Event:     make(chan BridgeEvent),
		Options:   options,

		SendFs: send_modfs,
		RecvFs: recv_modfs,

		syncLocalFrame:      1,
		syncRemoteFrame:     0,
		syncRemoteAckFrame:  0,
		syncLastRemoteFrame: 0,
		syncTimeoutCounter:  0,
	}
}

func (b *Bridge) event(e BridgeEvent) {
	select {
	case b.Event <- e:
	default:
	}
}

func (b *Bridge) connect() {
	if b.Connected {
		return
	}
	b.Connected = true
	b.event(BridgeConnect)
}

func (b *Bridge) disconnect() {
	if !b.Connected {
		return
	}
	b.Connected = false
	b.event(BridgeDisconnect)
}

func (b *Bridge) poll() bool {
	b.RecvFs.Read(false)

	syncFile, err := b.RecvFs.Get("sync")
	if err != nil {
		return false
	}

	lastActive := b.Connected
	lastRemoteFrame := b.syncRemoteFrame

	syncFile.Cursor = 0

	remoteVersion, _ := syncFile.ReadUint16()
	if remoteVersion != BRIDGE_VERSION {
		return false
	}

	b.syncRemoteFrame, _ = syncFile.ReadUint32()
	b.syncRemoteAckFrame, _ = syncFile.ReadUint32()

	ackFrameValid := b.syncRemoteAckFrame <= b.syncLocalFrame
	ackFrameThreshold := b.syncLocalFrame-b.syncRemoteAckFrame < 6

	b.syncLocalFrame++
	b.syncLastRemoteFrame = lastRemoteFrame

	if b.syncRemoteFrame > lastRemoteFrame {
		// active means coop is running and acknowledging us
		shouldActivate := ackFrameValid && ackFrameThreshold
		if shouldActivate {
			b.syncTimeoutCounter = 0
			if !lastActive {
				b.connect()
				log.Printf("Bridge connected\n")
			}
		}
		return b.Connected
	}

	if lastActive && !(ackFrameValid && ackFrameThreshold) {
		b.syncTimeoutCounter++
		if b.syncTimeoutCounter > 6 {
			log.Printf("Bridge disconnected - av:%t aft:%t slf:%d srf:%d sraf:%d stc:%d\n", ackFrameValid, ackFrameThreshold, b.syncLocalFrame, b.syncRemoteFrame, b.syncRemoteAckFrame, b.syncTimeoutCounter)
			b.disconnect()
		}
	}

	return false
}

func (b *Bridge) update() {
	if b.poll() {
	}

	syncFile := b.SendFs.Create("sync")
	syncFile.WriteUint16(BRIDGE_VERSION)
	syncFile.WriteUint32(b.syncLocalFrame)
	syncFile.WriteUint32(b.syncRemoteFrame)

	b.SendFs.Write()
}

func (b *Bridge) Start() {
	if b.Running {
		return
	}

	log.Println("Bridge running")

	b.Running = true

	for b.Running {
		b.update()
		interval := POLL_INTERVAL * time.Millisecond
		if !b.Connected {
			interval = POLL_INTERVAL_INACTIVE * time.Millisecond
		}
		time.Sleep(interval)
	}
}

func (b *Bridge) Stop() {
	if !b.Running {
		return
	}

	log.Println("Bridge stopping")

	b.Running = false
	b.Connected = false
}
